catalog/secrets/cache/src/main/java/org/projectnessie/catalog/secrets/cache/CachingSecretsBackend.java

```diff
@@ -0,0 +1,214 @@
+  private final LongSupplier clock;
+
+  public CachingSecretsBackend(SecretsCacheConfig config) {
+    OptionalLong ttl = config.ttlMillis();
+    this.ttlNanos = ttl.isPresent() ? TimeUnit.MILLISECONDS.toNanos(ttl.getAsLong()) : 0L;
+    this.clock = config.clockNanos();
+
+    Caffeine<CacheKeyValue, Secret> cacheBuilder =
+        Caffeine.newBuilder()
+            .expireAfter(
+                new Expiry<CacheKeyValue, Secret>() {
+                  @Override
+                  public long expireAfterCreate(
+                      CacheKeyValue key, Secret value, long currentTimeNanos) {
+                    long expire = key.expiresAtNanosEpoch;
+                    if (expire == NOT_CACHED) {
+                      return 0L;
+                    }
+                    long remaining = expire - currentTimeNanos;
+                    return Math.max(0L, remaining);
+                  }
+
+                  @Override
+                  public long expireAfterUpdate(
+                      CacheKeyValue key,
+                      Secret value,
+                      long currentTimeNanos,
+                      @NonNegative long currentDurationNanos) {
+                    return currentDurationNanos;
+                  }
+
+                  @Override
+                  public long expireAfterRead(
+                      CacheKeyValue key,
+                      Secret value,
+                      long currentTimeNanos,
+                      @NonNegative long currentDurationNanos) {
+                    return currentDurationNanos;
+                  }
+                })
+            .ticker(clock::getAsLong)
+            .maximumSize(config.maxElements());
+    config
+        .meterRegistry()
+        .ifPresent(
+            meterRegistry -> {
+              cacheBuilder.recordStats(() -> new CaffeineStatsCounter(meterRegistry, CACHE_NAME));
+              meterRegistry.gauge(
+                  "cache_max_size",
+                  singletonList(Tag.of("cache", CACHE_NAME)),
+                  "",
+                  x -> config.maxElements());
+              meterRegistry.gauge(
+                  "cache_element_ttl",
+                  singletonList(Tag.of("cache", CACHE_NAME)),
+                  "",
+                  x -> config.ttlMillis().orElse(0));
+            });
+
+    this.cache = cacheBuilder.build();
+  }
+
+  <S extends Secret> Optional<S> resolveSecret(
+      String repositoryId,
+      SecretsProvider backend,
+      URI name,
+      SecretType secretType,
+      Class<S> secretJavaType) {
+    long ttl = ttlNanos;
+    long expires = ttl != 0L ? clock.getAsLong() + ttl : 0L;
+
+    CacheKeyValue key = new CacheKeyValue(repositoryId, name, expires);
+
+    Secret fromCache =
+        cache.get(
+            key,
+            k -> {
+              @SuppressWarnings("unchecked")
+              Optional<Secret> loaded =
+                  (Optional<Secret>) backend.getSecret(name, secretType, secretJavaType);
```

Comment:
`backend.getSecret()` can block... Are we ok with that?

Reply:
yes